url0d38a10ee7_diegopliebana_GEBT_tgz-pJ8-ch_idsia_scenarios_test_EvolveWithChangingSeedsJ8/src/ch/idsia/mario/engine/level/Level.java

```diff
@@ -0,0 +1,252 @@
+        if (header != Level.FILE_HEADER) throw new IOException("Bad level header");
+        int version = dis.read() & 0xff;
+//        System.out.println("version = " + version);
+        int width = dis.readShort() & 0xffff;
+        int height = dis.readShort() & 0xffff;
+        Level level = new Level(width, height);
+        level.map = new byte[width][height];
+        level.data = new byte[width][height];
+        for (int i = 0; i < width; i++)
+        {
+            dis.readFully(level.map[i]);
+            dis.readFully(level.data[i]);
+        }
+        return level;
+    }
+
+    public void save(DataOutputStream dos) throws IOException
+    {
+        dos.writeLong(Level.FILE_HEADER);
+        dos.write((byte) 0);
+
+        dos.writeShort((short) width);
+        dos.writeShort((short) height);
+
+        for (int i = 0; i < width; i++)
+        {
+            dos.write(map[i]);
+            dos.write(data[i]);
+        }
+    }
+
+    public void tick()
+    {
+        for (int x = 0; x < width; x++)
+        {
+            for (int y = 0; y < height; y++)
+            {
+                if (data[x][y] > 0) data[x][y]--;
+            }
+        }
+    }
+
+    public byte getBlockCapped(int x, int y)
+    {
+        if (x < 0) x = 0;
+        if (y < 0) y = 0;
+        if (x >= width) x = width - 1;
+        if (y >= height) y = height - 1;
+        return map[x][y];
+    }
+
+    public byte getBlock(int x, int y)
+    {
+        if (x < 0) x = 0;
+        if (y < 0) return 0;
+        if (x >= width) x = width - 1;
+        if (y >= height) y = height - 1;
+        return map[x][y];
+    }
+
+    public void setBlock(int x, int y, byte b)
+    {
+        if (x < 0) return;
+        if (y < 0) return;
+        if (x >= width) return;
+        if (y >= height) return;
+        map[x][y] = b;
+    }
+
+    public void setBlockData(int x, int y, byte b)
+    {
+        if (x < 0) return;
+        if (y < 0) return;
+        if (x >= width) return;
+        if (y >= height) return;
+        data[x][y] = b;
+    }
+
+    public boolean isBlocking(int x, int y, float xa, float ya)
+    {
+        byte block = getBlock(x, y);
+        boolean blocking = ((TILE_BEHAVIORS[block & 0xff]) & BIT_BLOCK_ALL) > 0;
+        blocking |= (ya > 0) && ((TILE_BEHAVIORS[block & 0xff]) & BIT_BLOCK_UPPER) > 0;
+        blocking |= (ya < 0) && ((TILE_BEHAVIORS[block & 0xff]) & BIT_BLOCK_LOWER) > 0;
+
+        return blocking;
+    }
+
+    public SpriteTemplate getSpriteTemplate(int x, int y)
+    {
+        if (x < 0) return null;
+        if (y < 0) return null;
+        if (x >= width) return null;
+        if (y >= height) return null;
+        return spriteTemplates[x][y];
+    }
+
+    public void setSpriteTemplate(int x, int y, SpriteTemplate spriteTemplate)
+    {
+        if (x < 0) return;
+        if (y < 0) return;
+        if (x >= width) return;
+        if (y >= height) return;
+        spriteTemplates[x][y] = spriteTemplate;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;spriteTemplate&#39; (index: 2 | type: SpriteTemplate). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;x&#39;, &#39;y&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).